vendors/nuvoton/boards/numaker_iot_m487_wifi/aws_demos/application_code/nuvoton_code/entropy_hardware_poll.c

```diff
@@ -0,0 +1,221 @@
+    adc_val[oldest] = new_val;
+    oldest = (oldest + 1) % SNUM;
+
+    return ret_val;
+}
+
+
+uint32_t  adc_trng_gen_rnd()
+{
+    int       i;
+    uint32_t  val32;
+
+    val32 = 0;
+    for (i = 31; i >= 0; i--)
+        val32 |= (adc_trng_gen_bit() << i);
+    //printf("### %s: 0x%x \n", __FUNCTION__, val32);
+
+    return val32;
+}
+
+void  init_adc_init()
+{
+    static uint8_t init_flag = FALSE;
+    int    i;
+    
+    if( init_flag ) return;
+  
+    /* ADC refernce external 1.2V */
+    /* Unlock protected registers */
+    //SYS_UnlockReg();    
+    //SYS->VREFCTL = SYS_VREFCTL_VREF_PIN;
+    /* Lock protected registers */
+    //SYS_LockReg();
+  
+    /* Enable EADC clock */
+    CLK->APBCLK0 |= CLK_APBCLK0_EADCCKEN_Msk;
+
+    /* Set EADC clock divider */
+    CLK->CLKDIV0 &= ~CLK_CLKDIV0_EADCDIV_Msk;
+    CLK->CLKDIV0 |= (5 << CLK_CLKDIV0_EADCDIV_Pos);
+
+    EADC->CTL = (0x3 << EADC_CTL_RESSEL_Pos) | EADC_CTL_ADCEN_Msk;        /* A/D Converter Enable, select 12-bit ADC result  */
+
+    while (!(EADC->PWRM & EADC_PWRM_PWUPRDY_Msk));
+
+    EADC->SCTL[16] = (0x70 << EADC_SCTL_EXTSMPT_Pos)  /* ADC Sampling Time Extend          */
+                     | (0x0 << EADC_SCTL_TRGSEL_Pos);      /* A/D SAMPLE Start of Conversion Trigger Source Selection */
+
+    val_sum = 0;
+    for (i = 0; i < SNUM; i++)
+    {
+        adc_val[i] = get_adc_bg_val();
+        val_sum += adc_val[i];
+    }
+    oldest = 0;
+    init_flag = TRUE;
+    adc_trng_gen_rnd();    // drop the first 32-bits
+}
+
+void CRYPTO_IRQHandler()
+{
+    if (PRNG_GET_INT_FLAG(CRPT)) {
+        g_PRNG_done = 1;
+        PRNG_CLR_INT_FLAG(CRPT);
+    } else	if (AES_GET_INT_FLAG(CRPT)) {
+        g_AES_done = 1;
+        AES_CLR_INT_FLAG(CRPT);
+    }
+
+} 
+
+static void trng_get(unsigned char *pConversionData)
+{
+	uint32_t *p32ConversionData;
+    uint32_t u32val;
+  
+	p32ConversionData = (uint32_t *)pConversionData;
+	
+    /* Unlock protected registers */
+    SYS_UnlockReg();	
+    /* Enable IP clock */
+    CLK_EnableModuleClock(CRPT_MODULE);
+	
+    /* Lock protected registers */
+    SYS_LockReg();	
+	
+    NVIC_EnableIRQ(CRPT_IRQn);
+    PRNG_ENABLE_INT(CRPT);
+	
+    u32val = adc_trng_gen_rnd();
+    //printf("=== %s: 0x%x \n", __FUNCTION__, u32val);
+    PRNG_Open(CRPT, PRNG_KEY_SIZE_256, 1, u32val); //adc_trng_gen_rnd());
+
+    PRNG_Start(CRPT);
+    while (!g_PRNG_done);
+
+    PRNG_Read(CRPT, p32ConversionData);
+
+//    printf("    0x%08x  0x%08x  0x%08x  0x%08x\n\r", *p32ConversionData, *(p32ConversionData+1), *(p32ConversionData+2), *(p32ConversionData+3));
+//    printf("    0x%08x  0x%08x  0x%08x  0x%08x\n\r", *(p32ConversionData+4), *(p32ConversionData+5), *(p32ConversionData+6), *(p32ConversionData+7));
+
+    PRNG_DISABLE_INT(CRPT);
+///    NVIC_DisableIRQ(CRPT_IRQn);
+ //    CLK_DisableModuleClock(CRPT_MODULE);
+		
+}
+
+
+/*
+ * Get len bytes of entropy from the hardware RNG.
+ */
+ 
+int mbedtls_hardware_poll( void *data,
```

Comment:
Are we still going to leave a disclaimer comment about using a truly random number generator (looks like the ADC is used here)?

Reply:
@sarenameas , ADC is really used with one simple algorithm of white noise to generate one random seed for PRNG. However, not guarantee to get TRNG certification. How about your option to revise this comment ?